File: Python/Game.py
from Board import Board
from BoardView import BoardView
from UserInterface import UserInterface
from Player import Player
from Human import Human
from Computer import Computer
import random
import sys

class Game:
    _dicerolls = []

    """""""""""""""""""""""""""""""""""""""
     Function Name: __init__
     Purpose: constructor
     Parameters: None
     Return Value: None
     Local Variables: None
     Algorithm:
     Assistance Recieved: None
     """""""""""""""""""""""""""""""""""""""
    def __init__(self):
        self._firstmove = ""
        self._currentturn = ""

    """""""""""""""""""""""""""""""""""""""
     Function Name: firstmove
     Purpose: property (getter)
     Parameters: None
     Return Value: firstmove
     Local Variables: None
     Algorithm:
     Assistance Recieved: None
     """""""""""""""""""""""""""""""""""""""
    @property
    def firstmove(self):
        return self._firstmove

    """""""""""""""""""""""""""""""""""""""
     Function Name: currentturn
     Purpose: property (getter)
     Parameters: None
     Return Value: currentturn
     Local Variables: None
     Algorithm:
     Assistance Recieved: None
     """""""""""""""""""""""""""""""""""""""
    @property
    def currentturn(self):
        return self._currentturn

    """""""""""""""""""""""""""""""""""""""
     Function Name: dicerolls
     Purpose: property (getter)
     Parameters: None
     Return Value: dicerolls, list of ints
     Local Variables: None
     Algorithm:
     Assistance Recieved: None
     """""""""""""""""""""""""""""""""""""""
    @property
    def dicerolls(self):
        return self._dicerolls

    """""""""""""""""""""""""""""""""""""""
     Function Name: firstmove
     Purpose: property (setter)
     Parameters: player, string
     Return Value: None
     Local Variables: None
     Algorithm:
     Assistance Recieved: None
     """""""""""""""""""""""""""""""""""""""
    @firstmove.setter
    def firstmove(self, player):
        self._firstmove = player

    """""""""""""""""""""""""""""""""""""""
     Function Name: currentturn
     Purpose: property (setter)
     Parameters: player, string
     Return Value: None
     Local Variables: None
     Algorithm:
     Assistance Recieved: None
     """""""""""""""""""""""""""""""""""""""
    @currentturn.setter
    def currentturn(self, player):
        self._currentturn = player

    """""""""""""""""""""""""""""""""""""""
     Function Name: runGame
     Purpose: execute a game
     Parameters:
                handicap, a dictionary holding info about the handicap
                scores, a dictionary holding the scores of each player
                filename, string of file to read from
     Return Value: None
     Local Variables: None
     Algorithm:
            check if file needs to be loaded
                load game
                check if loaded game has been won
                if so declare winner
            ask for size
            create a board
            figure out who goes first
            draw the board
            enter loop
            execute a turn
            ask to save (if yes, save and exit)
            if game was not won switch player
            exit loop when game is won
            calculate score
            declare winner
     Assistance Recieved: None
     """""""""""""""""""""""""""""""""""""""
    def runGame(self, handicap, scores, filename):
        if filename == "":
            board = Board(UserInterface.askSize(), handicap)
            self.rollForFirst()
        else:
            board = self.copyFromFile(filename + ".txt")
            if board.isGameWon():
                BoardView.drawBoard( board )
                switchTurn( board )
                points = calcScore( board )
                UserInterface.declareGameWon(self.currentturn, points)
                return (self.currentturn, self.firstmove, points)
        BoardView.drawBoard( board )
        while True:
            UserInterface.declareTurn(self.currentturn)
            self.startTurn( board )
            if UserInterface.askToSave():
                self.saveGame(board, scores)
                sys.exit()
            if board.isGameWon():
                break
            else:
                self.switchTurn(board)
                BoardView.drawBoard(board)
        points = self.calcScore( board )
        UserInterface.declareGameWon( self.currentturn, points )
        return {"winner" : self.currentturn, "firstmove" : self.firstmove, "points" : points}

    """""""""""""""""""""""""""""""""""""""
     Function Name: rollForFirst
     Purpose: decide which player goes first
     Parameters: None
     Return Value: None
     Local Variables: None
     Algorithm:
     Assistance Recieved: None
     """""""""""""""""""""""""""""""""""""""
    def rollForFirst(self):
        UserInterface.playersRoll()
        humanroll = 0
        comroll = 0
        while humanroll == comroll:
            humanroll = UserInterface.rollResult( "Human", self.rollDie(), self.rollDie() )
            comroll = UserInterface.rollResult( "Computer", self.rollDie(), self.rollDie() )
        if humanroll > comroll:
            self.firstmove = "Human"
        else:
            self.firstmove = "Computer"
        self.currentturn = self.firstmove
        UserInterface.declareFirst(self.firstmove)

    """""""""""""""""""""""""""""""""""""""
     Function Name: canPlay
     Purpose: roll a random die or read die from list
     Parameters: None
     Return Value: int, the roll
     Local Variables: None
     Algorithm:
     Assistance Recieved: None
     """""""""""""""""""""""""""""""""""""""
    def rollDie(self):
        if self.dicerolls:
            roll = self.dicerolls[0]
            self.dicerolls.remove(roll)
            return roll
        else:
            return random.randint(1, 6)

    """""""""""""""""""""""""""""""""""""""
     Function Name: startTurn
     Purpose: execute a turn
     Parameters:
                Board, tis game's board
     Return Value: None
     Local Variables: None
     Algorithm:
        in loop
            decide how many dice to roll
            roll for move
            check if moves available
            decide to cover or uncover
            choose squares to play
            update board
        exit when game is won or no moves are available
        draw the board
     Assistance Recieved: None
     """""""""""""""""""""""""""""""""""""""
    def startTurn(self, board):
        if self.currentturn == "Human":
            player = Human()
        else:
            player = Computer()
        while True:
            if player.canThrowOne(board):
                numdice = player.chooseDice(board)
            else:
                numdice = 2
            if numdice == 2:
                roll = UserInterface.rollResult( self.currentturn, self.rollDie(), self.rollDie() )
            else:
                roll = UserInterface.rollResult( self.currentturn, self.rollDie(), 0)
            if player.canPlay(board, roll):
                coverself = player.decideCover(board, roll)
                squares = player.chooseSquares(board, coverself, roll)
                if self.currentturn == "Computer":
                    UserInterface.computerMove(coverself, squares)
                board.updateBoard(squares, self.currentturn, coverself)
                BoardView.drawBoard(board)
            else:
                UserInterface.noMoves()
                return
            if board.isGameWon():
                return

    """""""""""""""""""""""""""""""""""""""
     Function Name: switchTurn
     Purpose: switch players
     Parameters: board, the game's board
     Return Value: None
     Local Variables: None
     Algorithm:
     Assistance Recieved: None
     """""""""""""""""""""""""""""""""""""""
    def switchTurn(self, board):
        if self.currentturn == "Human":
            self.currentturn = "Computer"
        else:
            self.currentturn = "Human"
        board.setTurnFlag()

    """""""""""""""""""""""""""""""""""""""
     Function Name: calcScore
     Purpose: calculate the winner's score
     Parameters: None
     Return Value: board, this game's board
     Local Variables: None
     Algorithm:
     Assistance Recieved: None
     """""""""""""""""""""""""""""""""""""""
    def calcScore(self, board):
        test = board.comboard[1]
        for square in board.comboard:
            if board.comboard[square] != test:
                return self.addSquares( board.comboard, not board.humanboard[1] )
        return self.addSquares( board.humanboard, not test )

    """""""""""""""""""""""""""""""""""""""
     Function Name: addSquares
     Purpose: add up the squares to determine the winner's score
     Parameters:
                side, which side to add up
                covered, boolean value to compare to
     Return Value: int, the total
     Local Variables: None
     Algorithm:
     Assistance Recieved: None
     """""""""""""""""""""""""""""""""""""""
    def addSquares(self, side, covered):
        total = 0
        for square in side:
            if side[square] == covered:
                total += square
        return total

    """""""""""""""""""""""""""""""""""""""
     Function Name: saveGame
     Purpose: write game info to text file
     Parameters:
                Board, this game's board
                scores, dict containing each player's score
     Return Value: None
     Local Variables: None
     Algorithm:
     Assistance Recieved: None
     """""""""""""""""""""""""""""""""""""""
    def saveGame(self, board, scores):
        self.switchTurn(board)
        filename = UserInterface.filePrompt()
        f = open(filename + ".txt", 'w')
        f.write("Computer:\n\tSquares: ")
        for square in range(1, board.boardsize + 1):
            if board.comboard[square]:
                f.write("* ")
            else:
                f.write(str(square) + " ")
        f.write("\n\tScore: " + str(scores["computer"]) + "\n\nHuman:\n\tSquares: ")
        for square in range(1, board.boardsize + 1):
            if board.humanboard[square]:
                f.write("* ")
            else:
                f.write(str(square) + " ")
        f.write("\n\tScore: " + str(scores["human"]) )
        f.write("\n\nFirst Turn: " + self.firstmove + "\nNext Turn: " + self.currentturn)
        f.close()
        print("Game Saved.\n")

    """""""""""""""""""""""""""""""""""""""
     Function Name: copyFromFile
     Purpose: read game info from file
     Parameters: filename, string of file to read from
     Return Value: the game's board
     Local Variables: None
     Algorithm:
     Assistance Recieved: None
     """""""""""""""""""""""""""""""""""""""
    def copyFromFile(self, filename):
        f = open(filename, 'r')
        for line in f:
            tokens = line.split()
            if tokens:
                if tokens[0] == "Computer:":
                    line = f.readline()
                    tokens = line.split()
                    tokens.remove("Squares:")
                    board = Board(len(tokens), {})
                    for square in range(0, len(tokens)):
                        if tokens[square] == "*":
                            board.setSquare(board.comboard, square+1, True)
                if tokens[0] == "Human:":
                    line = f.readline()
                    tokens = line.split()
                    tokens.remove("Squares:")
                    for square in range(0, len(tokens)):
                        if tokens[square] == "*":
                            board.setSquare(board.humanboard, square+1, True)
                if tokens[0] == "First":
                    self.firstmove = tokens[2]
                if tokens[0] == "Next":
                    self.currentturn = tokens[2]
                if tokens[0] == "Dice:":
                    while True:
                        line = f.readline()
                        if line:
                            rolls = line.split()
                            for roll in rolls:
                                self.dicerolls.append( int(roll) )
                        else:
                            break
        f.close()
        board.setTurnFlag()
        return board

File: C++/implementation/Game.cpp
#include "Game.h"
#include "Turn.h"
#include "BoardView.h"
#include "UserInterface.h"
#include <iostream>
#include <fstream>
#include <sstream>
using namespace std;

/*************************
 Function Name: Game
 Purpose: Class default constructor. Initialize member variables.
 Parameters: None
 Return Value: None
 Local Variables: None
 Algorithm: Initialize strings to "N/A"
 Assistance Recieved: None
 *************************/
Game::Game()
{
    firstmove = "N/A";
    currentturn = "N/A";
}

/*************************
 Function Name: runGame
 Purpose: execute a game
 Parameters: 
            Handicap, a pointer to a Handicap structure
            Tourn, Tournament object used to access scores when saving
            filename, string of file to load
 Return Value: 
            Winner, winner structure containing neccessary information about the game's outcome
 Local Variables: 
            Board, the game's board
            temp, a temporary board used to initialize a board if no gameboard is being loaded
            points, an int to hold the points earned by the winner of a game
 Algorithm:
            check if file needs to be loaded
                check if loaded game has been won
                if so declare winner
            ask for size
            create a board
            figure out who goes first
            draw the board
            enter loop
            execute a turn
            ask to save (if yes, save and exit)
            if game was not won switch player
            exit loop when game is won
            calculate score
            declare winner
 Assistance Recieved: None
 *************************/
Winner Game::runGame(Handicap* Handicap, Tournament Tourn, string filename)
{
    Board Board;
    //check if file needs to be loaded
    if (filename == "") {
        //new game, get size of board from user and create the board
        class Board temp( UserInterface::askSize(), Handicap );
        Board = temp;
        //decide who goes first
        rollForFirst();
    }
    else
    {
        //game needs to be loaded
        Board = copyFromFile(filename);
        if (Board.isGameWon()) {
            //game was won
            BoardView::drawBoard( Board );
            //set current turn to the winner
            switchTurn(Board);
            int points = calcScore( Board );
            UserInterface::declareGameWon( getCurrentTurn(), points );
            return Winner( getCurrentTurn(), getFirstMove(), points );
        }
    }
    BoardView::drawBoard( Board );
    do
    {
        Turn::startTurn( Board, getCurrentTurn() );
        if ( UserInterface::askToSave() )
        {
            //save game
            //set the current turn to the next player to go
            switchTurn(Board);
            saveGame(Board, Tourn);
            exit(0);
        }
        if ( !Board.isGameWon() )
        {
            //game not won
            switchTurn(Board);
            BoardView::drawBoard( Board );
        }
    } while ( !Board.isGameWon() );
    //game won
    int points = calcScore( Board );
    UserInterface::declareGameWon( getCurrentTurn(), points );
    return Winner( getCurrentTurn(), getFirstMove(), points );
}

/*************************
 Function Name: getFirstMove
 Purpose: gets which player went first;
 Parameters: None
 Return Value: a string, the player that went first
 Local Variables: None
 Algorithm:
            return the member variable
 Assistance Recieved: None
 *************************/
string Game::getFirstMove() const
{
    return firstmove;
}

/*************************
 Function Name: getCurrentTurn
 Purpose: get whose turn it is
 Parameters: None
 Return Value: a string, the player whose turn it is
 Local Variables: None
 Algorithm:
            return the member variable
 Assistance Recieved: None
 *************************/
string Game::getCurrentTurn() const
{
    return currentturn;
}

/*************************
 Function Name: setFirstMove
 Purpose: changes value of variable firstmove
 Parameters: string, who goes first
 Return Value: None
 Local Variables:
 Algorithm:
 set firstmove to the passed string
 Assistance Recieved: None
 *************************/
void Game::setFirstMove( string x )
{
    firstmove = x;
}

/*************************
 Function Name: setCurrentTurn
 Purpose: changes value of variable currentturn
 Parameters: string, whose turn it is
 Return Value: None
 Local Variables:
 Algorithm:
 set currentturn to the passed string
 Assistance Recieved: None
 *************************/
void Game::setCurrentTurn( string x )
{
    currentturn = x;
}

/*************************
 Function Name: rollForFirst
 Purpose: roll dice to decide who has first move
 Parameters: None
 Return Value: None
 Local Variables:
                humanroll, an int to hold human's total dice roll
                comroll, an int to hold com's total dice roll
 Algorithm:
            tell user players will roll to decide whose first
            enter loop
            roll dice until one is higher
            set the first player
 Assistance Recieved: None
 *************************/
void Game::rollForFirst()
{
    UserInterface::playersRoll();
    int humanroll;
    int comroll;
    do
    {
        humanroll = UserInterface::rollResult("Human", Game::rollDie(), Game::rollDie());
        comroll = UserInterface::rollResult("Computer", Game::rollDie(), Game::rollDie());
    } while ( humanroll == comroll );
    setFirstMove( decideFirst(humanroll, comroll) );
    setCurrentTurn( getFirstMove() );
}

/*************************
 Function Name: decideFirst
 Purpose: decides which player goes first each game
 Parameters: 
            humanroll, an int to hold human's total dice roll
            comroll, an int to hold com's total dice roll
 Return Value: string, the player with the higher roll
 Local Variables: None
 Algorithm:
            compare rolls and assign firstmove corresponding value
            return player with higher roll
 Assistance Recieved: None
 *************************/
string Game::decideFirst(int humanroll, int comroll)
{
    if ( humanroll > comroll )
    {
        cout << "Human will go first.\n";
        return "Human";
    }
    else if ( comroll > humanroll )
    {
        cout << "Computer will go first.\n";
        return "Computer";
    }
    return "";
}

/*************************
 Function Name: rollDie
 Purpose: emulate the rolling of a die
 Parameters: None
 Return Value: an int 1-6, the roll of a die
 Local Variables: None
 Algorithm:
 create random number between 1-6
 return that value
 Assistance Recieved: None
 *************************/
int Game::rollDie()
{
    /* use this for testing
    int roll;
    cout << "Enter roll: ";
    cin >> roll;
    return roll;
     */
    
    return (rand() % 6) + 1;
}

/*************************
 Function Name: switchTurn
 Purpose: switch the current turn to the next player
 Parameters:
            Board, passed by reference to set the turnflag
 Return Value: void
 Local Variables: None
 Algorithm:
 Assistance Recieved: None
 *************************/
void Game::switchTurn(Board &Board)
{
    if ( getCurrentTurn() != getFirstMove() ) {
        Board.setTurnFlag();
    }
    if ( getCurrentTurn() == "Human" )
    {
        setCurrentTurn("Computer");
    }
    else
    {
        setCurrentTurn("Human");
    }
}

/*************************
 Function Name: saveGame
 Purpose: save the game
 Parameters:
 Return Value:
 Local Variables:
 Algorithm:
 Assistance Recieved: None
 *************************/
void Game::saveGame(Board Board, Tournament Tourn)
{
    string filename;
    cout << "Name the save file: ";
    cin >> filename;
    filename += ".txt";
    
    ofstream file;
    file.open(filename);
    file << "Computer: \n\tSquares: ";
    for (int i = 0; i < Board.getBoardSize(); i++) {
        if ( Board.testSquare( i , Board.getComBoard(), true ) )
        {
            file << "* ";
        }
        else
        {
            file << i+1 << " ";
        }
    }
    file << "\n\tScore: " << Tourn.getComScore();
    file << "\n\nHuman: \n\tSquares: ";
    for (int i = 0; i < Board.getBoardSize(); i++) {
        if ( Board.testSquare( i , Board.getHumanBoard(), true ) )
        {
            file << "* ";
        }
        else
        {
            file << i+1 << " ";
        }
    }
    file << "\n\tScore: " << Tourn.getHumanScore();
    file << "\n\nFirst Turn: " << getFirstMove();
    file << "\nNext Turn: " << getCurrentTurn();
    file.close();
    cout << "Game saved as " << filename << endl;
}

int Game::calcScore(Board Board)
{
    bool test = Board.getComBoard()[0];
    for (int i = 1; i < Board.getBoardSize(); i++) {
        if ( Board.getComBoard()[i] != test )
        {
            return addsquares( Board.getComBoard(), Board.getBoardSize(), !Board.getHumanBoard()[0] );
        }
    }
    return addsquares( Board.getHumanBoard(), Board.getBoardSize(), !test );
}

int Game::addsquares(vector<bool> board, int size, bool covered )
{
    int sum = 0;
    for (int i = 0; i < size; i++) {
        if ( board[i] == covered )
        {
            sum += i + 1;
        }
    }
    return sum;
}

Board Game::copyFromFile(string filename)
{
    int size = 0;
    vector<int> comsquares;
    vector<int> humansquares;
    string firstmove;
    string currentturn;
    string input;
    ifstream file;
    file.open(filename);
    if (file.is_open())
    {
        string line;
        while ( getline(file, line) )
        {
            if ( line.find("Computer:", 0) != string::npos )
            {
                getline(file, line);
                istringstream ss(line);
                ss >> input;
                int square = 1;
                while(ss >> input)
                {
                    if (input == "*") {
                        comsquares.push_back(square);
                    }
                    size++;
                    square++;
                }
            }
            if ( line.find("Human:", 0) != string::npos )
            {
                getline(file, line);
                istringstream ss(line);
                ss >> input;
                int square = 1;
                while(ss >> input)
                {
                    if (input == "*") {
                        humansquares.push_back(square);
                    }
                    square++;
                }
            }
            if ( line.find("First ", 0) != string::npos )
            {
                istringstream ss(line);
                ss >> firstmove;
                ss >> firstmove;
                ss >> firstmove;
            }
            if ( line.find("Next ", 0) != string::npos )
            {
                istringstream ss(line);
                ss >> currentturn;
                ss >> currentturn;
                ss >> currentturn;
            }
        }
    }
    Board Board( size );
    Board.updateBoard(comsquares, "Computer", true);
    Board.updateBoard(humansquares, "Human", true);
    Board.setTurnFlag();
    setFirstMove(firstmove);
    setCurrentTurn(currentturn);
    return Board;
}


